src/middleware/FileRestrictionsMiddleware.ts

```diff
@@ -0,0 +1,94 @@
+import { APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { Attachment } from 'app/models/Attachment';
+
+const customErrorTemplate = 'error-custom';
+
+@provide(FileRestrictionsMiddleware)
+export class FileRestrictionsMiddleware extends BaseMiddleware {
+
+    public handler(req: Request, res: Response, next: NextFunction): void {
+
+        const session = req.session;
+
+        const adminPermissionFlag = session.chain(_ => _.getValue<ISignInInfo>(SessionKey.SignInInfo))
+            .chainNullable(signInInfo => signInInfo[SignInInfoKeys.AdminPermissions])
+            .extract();
+
+        const userProfile: IUserProfile = session.chain(_ => _.getValue<ISignInInfo>(SessionKey.SignInInfo))
+            .chainNullable(signInInfo => signInInfo[SignInInfoKeys.UserProfile])
+            .ifNothing(() => loggerInstance().error(`${FileRestrictionsMiddleware.name} - User profile was expected in session but none found`))
+            .unsafeCoerce();
+
+        const permissions = userProfile[UserProfileKeys.Permissions];
+
+        if (adminPermissionFlag === '1') {
+            if (permissions &&
+                permissions[AppealsPermissionKeys.download] &&
+                permissions[AppealsPermissionKeys.view]) {
+                return next();
+            }
+            return this.renderForbiddenError(res);
+
+        } else {
+
+            const appeal: Appeal = session.chain(_ => _.getExtraData())
+                .chainNullable(extraData => extraData[APPLICATION_DATA_KEY])
+                .chainNullable(applicationData => applicationData.appeal)
+                .ifNothing(() => loggerInstance().error(`${FileRestrictionsMiddleware.name} - Appeal was expected in session but none found`))
+                .unsafeCoerce();
+
+            const fileId = req.params.fileId;
+            const attachement = this.getAttachmentFrom(appeal, fileId);
+
+            if (!attachement) {
+                return this.renderForbiddenError(res);
+            }
+
+            // User must be creating a new appeal.
+            if (!appeal.createdBy) {
+                return next();
+            }
+
+            // Appeal was loaded from API
+            if (appeal.createdBy?._id !== userProfile.id) {
+
+                return this.renderForbiddenError(res);
+            }
+
+        }
+
+        return next();
+    }
+
+    private renderForbiddenError(res: Response): void {
+        res.status(FORBIDDEN);
+        res.render(customErrorTemplate, {
+            heading: 'You are not authorised to download this document'
+        });
+    }
+
+    private getAttachmentFrom(appeal: Appeal, fileId: string): Attachment | undefined {
+
+        if (!fileId) {
+            throw Error('File id must not be null');
+        }
+
+        if (appeal.reasons.other.attachments) {
+            const attachment = appeal.reasons.other.attachments.find(attachement => attachement.id === fileId);
```

Comment:
do we need this local variable?

Reply:
thought it would make it more readable :D

Reply:
this would be slightly more readable for me:

```
      const attachments: Attachment[] | undefined = appeal.reasons.other.attachments;

        if (attachments) {
            return attachments.find(attachment => attachment.id === fileId);
        }
```